executor: glibc pkey APIs break musl builds; glibc pthread_create + seccomp TRAP on clone3 breaks sandboxed fuzzing

### Description
  We are fuzzing Linux on **arm64 devices with an app-level seccomp sandbox** (similar to Android's application filter). Building and running the executor
   has exposed two related libc compatibility issues that did not exist before the pkey protection support was added.
  #### Background: pkey usage in the executor
  Since commit c3a6603be ("executor: protect kcov/output regions with pkeys"), `executor/executor_linux.h` uses glibc-specific memory protection key APIs:
  - `pkey_alloc()` / `pkey_free()` — probe for pkey support at startup
  - `pkey_mprotect()` — associate the KCOV buffer (and snapshot output buffer) with a reserved pkey
  - `pkey_set()` — temporarily disable/enable write access while reading coverage data
  This is guarded at runtime by a `pkeys_enabled` flag: if `pkey_alloc()` fails (e.g. kernel returns ENOSYS), all pkey calls are skipped. So on platforms 
  without pkey support, the feature is effectively a no-op at runtime.
  However, the **compile-time dependency on glibc headers** (`<sys/mman.h>` pkey declarations) prevents building the executor with **musl**, which we 
  previously used successfully as a static cross-compiler for arm64 targets.
  #### Problem 1: musl cross-compilation fails
  With musl (e.g. `aarch64-linux-musl-g++`), the build fails because musl does not provide `pkey_alloc`, `pkey_free`, `pkey_mprotect`, or `pkey_set` in 
  its headers:

  executor/executor_linux.h:84: error: 'pkey_alloc' was not declared in this scope
  executor/executor_linux.h:94: error: 'pkey_free' was not declared in this scope
  ...

  Before c3a6603be, musl-based static builds of `syz-executor` worked fine for our arm64 deployment.
  **Suggested fix:** Add a small compat layer (direct `syscall(__NR_pkey_*)` wrappers behind `#ifndef __GLIBC__` or feature detection), or move pkey calls
   behind `#ifdef` with stub implementations that set `pkeys_enabled = false`. Since the runtime path already handles ENOSYS gracefully, compile-time
  stubs would preserve existing behavior on platforms without pkeys.
  #### Problem 2: glibc + app seccomp → pthread_create killed by SIGSYS on clone3
  When built with glibc (≥ 2.34), `pthread_create()` internally prefers the `clone3(435)` syscall. Our app seccomp filter **TRAPs syscall 435** (while
  allowing legacy `clone(220)`).
  Observed behavior on arm64 with seccomp installed:
  | Test | Result |
  |------|--------|
  | Direct `clone(220)` | OK |
  | Direct `clone3(435)` | SIGSYS |
  | `pthread_create()` with glibc | SIGSYS (via clone3) |
  | `pthread_create()` with musl | OK (uses clone) |
  glibc only falls back from clone3 to clone when clone3 returns **ENOSYS**. A seccomp **TRAP** does not trigger this fallback — the process receives
  SIGSYS and dies.
  This affects any code path that creates threads after the sandbox seccomp filter is installed, including `thread_start()` in `executor/common.h` and the
   event thread in `common_linux.h`.
  Workarounds we have tested:
  1. **Skip installing the app seccomp filter** — fuzzing works, but loses the sandbox layer.
  2. **Use musl instead of glibc** — avoids clone3, but currently blocked by Problem 1.
  We have **not** found a `GLIBC_TUNABLES` setting to disable the clone3 code path in glibc's pthread implementation.
  #### Environment
  - Target: arm64 Linux (mobile/embedded device with app sandbox)
  - Host build: x86_64 Linux
  - Cross-compilers tested:
    - `aarch64-linux-gnu-g++` (glibc) — compiles, fails at runtime with seccomp
    - `aarch64-linux-musl-g++` (musl 1.2.x) — fails at compile time (pkey APIs)
  - syzkaller: recent master (post c3a6603be and readonly coverage support)
  - Sandbox: app-level seccomp filter installed via `SECCOMP_SET_MODE_FILTER`, TRAP on syscall 435
  #### Questions for maintainers
  1. Would upstream accept a pkey compat shim so musl (and other non-glibc) builds work again?
  2. Is there interest in a Harmony/Android-style sandbox code path that avoids glibc's clone3 dependency (e.g. optional `thread_start()` using
  `clone(220)` directly, or documenting that app seccomp + glibc is unsupported)?
  3. For platforms where pkeys are unavailable at runtime anyway, should the pkey code be fully optional at compile time?
  #### Possible solutions (for discussion)
  - **A.** `#ifndef __GLIBC__` pkey syscall stubs in `executor_linux.h` (minimal, preserves musl builds)
  - **B.** Optional executor build flag to disable pkey support entirely
  - **C.** Platform-specific `thread_start()` that uses `clone(220)` instead of `pthread_create()` when `SYZ_SANDBOX_*` with app seccomp is enabled
  - **D.** Document that glibc static builds require seccomp filters to allow clone3 or return ENOSYS (not TRAP) for syscall 435
  Happy to provide a PR for option A if there is interest.

--- a/executor/executor_linux.h
+++ b/executor/executor_linux.h
@@ -11,6 +11,40 @@
 #include <sys/syscall.h>
 #include <unistd.h>
 
+#ifndef __GLIBC__
+// musl (and other non-glibc libcs) don't declare the pkey_* wrappers or the
+// PKEY_DISABLE_WRITE flag, even though the kernel syscalls exist on any
+// modern Linux. Provide minimal equivalents so the executor keeps building there.
+// pkey_set() has no matching syscall (glibc implements it by writing the
+// PKRU register directly), so it's stubbed to report the feature as
+// unavailable; the callers below already treat that as "protection disabled",
+// same as when the kernel itself lacks pkey support.
+#ifndef PKEY_DISABLE_WRITE
+#define PKEY_DISABLE_WRITE 0x2
+#endif
+
+static int pkey_alloc(unsigned int flags, unsigned int access_rights)
+{
+	return syscall(__NR_pkey_alloc, flags, access_rights);
+}
+
+static int pkey_free(int pkey)
+{
+	return syscall(__NR_pkey_free, pkey);
+}
+
+static int pkey_mprotect(void* addr, size_t len, int prot, int pkey)
+{
+	return syscall(__NR_pkey_mprotect, addr, len, prot, pkey);
+}
+
+static int pkey_set(int pkey, unsigned int rights)
+{
+	errno = ENOSYS;
+	return -1;
+}
+#endif
+
 static bool pkeys_enabled;
 
 // The coverage buffer can realistically overflow. In the non-snapshot mode we cannot afford

--- a/executor/executor_linux.h
+++ b/executor/executor_linux.h
@@ -1,6 +1,7 @@
 // Copyright 2015 syzkaller project authors. All rights reserved.
 // Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
 
+#include <errno.h>
 #include <fcntl.h>
 #include <signal.h>
 #include <stdio.h>
@@ -25,17 +26,32 @@
 
 static int pkey_alloc(unsigned int flags, unsigned int access_rights)
 {
+#ifdef __NR_pkey_alloc
 	return syscall(__NR_pkey_alloc, flags, access_rights);
+#else
+	errno = ENOSYS;
+	return -1;
+#endif
 }
 
 static int pkey_free(int pkey)
 {
+#ifdef __NR_pkey_free
 	return syscall(__NR_pkey_free, pkey);
+#else
+	errno = ENOSYS;
+	return -1;
+#endif
 }
 
 static int pkey_mprotect(void* addr, size_t len, int prot, int pkey)
 {
+#ifdef __NR_pkey_mprotect
 	return syscall(__NR_pkey_mprotect, addr, len, prot, pkey);
+#else
+	errno = ENOSYS;
+	return -1;
+#endif
 }
 
 static int pkey_set(int pkey, unsigned int rights)

